python/cudf/cudf/core/column/string.py

```diff
@@ -4620,7 +4620,17 @@ def _string_column_binop(lhs, rhs, op, out_dtype):
     return out
 
 
-def _get_cols_list(others):
+def _get_cols_list(parent_obj, others):
+
+    if isinstance(parent_obj, cudf.Index):
+        # TODO: We might have to change
+        # this incase the fix of the below
+        # issue alters behaviour incase of
+        # Index object being parent_obj
+        # https://github.com/pandas-dev/pandas/issues/35556
+        parent_obj = parent_obj.to_series(
+            index=cudf.RangeIndex(len(parent_obj))
+        )
```

Comment:
If `others` are `Series` objects with a non-standard index won't this cause reindexing unnecessarily instead of treating this more like a column (without an index)?

Reply:
In such cases, we would still need to perform index alignment with the rest of the list of `Series` objects which have a non-standard index too. for example..

```python
>>> import cudf
>>> i = cudf.Index(["AbC", "de", "FGHI", "j", "kLm"])
>>> other = (
...             cudf.Series(["f", "g", "h", "i", "j"], index=['a','b','c','d','e']),
...             cudf.Series(["f", "g", "h", "i", "j"]),
... )
>>> print(i.str.cat(other))
StringIndex([None None None None None], dtype='object')

>>> i = cudf.Series(["AbC", "de", "FGHI", "j", "kLm"])
>>> print(i.str.cat(other))
0    <NA>
1    <NA>
2    <NA>
3    <NA>
4    <NA>
dtype: object
>>> 
```

Reply:
I guess the case I was trying to show:
```
import pandas as pd

pds = pd.Series(['a', 'b', 'c', 'd', 'e'], index=['a', 'b', 'c', 'd', 'e'])
str_list = ['f', 'g', 'h', 'i', 'j']
idx = pd.Index(str_list)

print(pds.str.cat(str_list))
# a    af
# b    bg
# c    ch
# d    di
# e    ej
# dtype: object

print(pds.str.cat(idx))
# This is the Pandas bug you're referencing
# a    NaN
# b    NaN
# c    NaN
# d    NaN
# e    NaN
# dtype: object
```

Where without the Pandas bug I'd expect the `idx` case above to equal the `str_list` case, and with your current implementation `idx` would get converted to a Series with index 0 --> N which would then be reindexed to `a` --> `e` and all the values would be null erroneously.

Reply:
Addressed this along with adding tests for Index.str.cat separately. Had to xfail some of them due to the issues in pandas and cudf. But majority of xfail's would go away with pandas 1.1.0 upgrade(as it has some fixes).

We could have one more optimization where we don't at all convert `parent_obj`(`Index`) to `Series`, if `others` contains no `Series` objects at all. But this would probably need a two-pass scan over `others`, where once we check for the existence of any `Series` objects and the second pass is to reindex.

Reply:
Update: Changed the code to extract the index based on the parent object, hence avoiding series creation overhead.